snuba/cli/bootstrap.py

```diff
@@ -100,11 +101,25 @@ def bootstrap(
     # tables or distributed tables, etc.
 
     # Create the tables for every dataset.
+    clickhouse_tables = ClickhousePool().execute("show tables")
+    existing_tables = [row[0] for row in clickhouse_tables]
     for name in DATASET_NAMES:
         dataset = get_dataset(name)
 
         logger.debug("Creating tables for dataset %s", name)
+        run_migrations = False
         for statement in dataset.get_dataset_schemas().get_create_statements():
-            logger.debug("Executing:\n%s", statement)
-            ClickhousePool().execute(statement)
+            if statement.table_name not in existing_tables:
+                # This hack is needed since ClickHouse would try to execute the
+                # SELECT statement that defines a Materialized View even if the View
+                # exists. This may require the migrate script to run instead to make
+                # the SELECT statement work.
+                logger.debug("Executing:\n%s", statement.statement)
+                ClickhousePool().execute(statement.statement)
+            else:
+                logger.debug("Skipping existing table %s", statement.table_name)
+                run_migrations = True
```

Comment:
It seems like there is the potential for this branch to only be hit on a subset of the total loop iterations, resulting in a mix of `CREATE` statements being run, and some being skipped, in which case the migrations are going to be run on the newly created table (and presumably do nothing.)

Would it make sense to just run the migrations block each time — even if they do nothing — just to reduce the branching factor here?

Reply:
It would not work.
Migrations are per dataset. One single list per dataset.
While the create statements in the loop are potentially multiple per schema per dataset. Specifically in the materialized views we have three create statements in one dataset and still one migration function. So we have to first run all the create statements for a dataset and then run the migration function for that dataset.